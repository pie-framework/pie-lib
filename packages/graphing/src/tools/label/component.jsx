import React from 'react';

export class Label extends React.Component {
  render() {
    return null;
  }
}

export default Label;
