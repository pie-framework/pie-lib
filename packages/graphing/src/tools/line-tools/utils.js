import { calculateThirdPointOnLine } from '../../utils';

export const arrowDimensions = {
  none: 0,
  vector: 7,
  ray: 5,
  line: 5
};

export const calculateCorrectScaledPoints = (scale, from, to, type = 'none') => {
  const scaledFromX = scale.x(from.x);
  const scaledFromY = scale.y(from.y);
  let scaledToX = scale.x(to.x);
  let scaledToY = scale.y(to.y);
  let middlePoint;

  middlePoint = calculateThirdPointOnLine(
    { x: scaledToX, y: scaledToY },
    { x: scaledFromX, y: scaledFromY },
    {
      domain: {
        min: scale.x(to.x) - arrowDimensions[type],
        max: scale.x(to.x) + arrowDimensions[type]
      },
      range: {
        min: scale.y(to.y) - arrowDimensions[type],
        max: scale.y(to.y) + arrowDimensions[type]
      }
    }
  );

  if (middlePoint && isFinite(middlePoint.x) && isFinite(middlePoint.y)) {
    scaledToX = middlePoint.x;
    scaledToY = middlePoint.y;
  }

  return {
    x: scaledToX,
    y: scaledToY
  };
};
