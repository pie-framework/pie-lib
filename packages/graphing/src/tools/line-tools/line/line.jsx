import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { types } from '@pie-lib/plot';
import { arrowDimensions, calculatePreviousNearestScaledPoint } from '../../../utils';

/**
 * A low level segment component
 */
export class RawLine extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    disabled: PropTypes.bool,
    from: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }),
    to: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }),
    forward: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }).isRequired,
    backward: PropTypes.shape({
      x: PropTypes.number,
      y: PropTypes.number
    }).isRequired,
    graphProps: types.GraphPropsType.isRequired
  };

  render() {
    const {
      classes,
      disabled,
      className,
      correctness,
      graphProps,
      backward,
      forward,
      ...rest
    } = this.props;
    const { scale } = graphProps;
    const { x: scaledBackwardX, y: scaledBackwardY } = calculatePreviousNearestScaledPoint(
      scale,
      forward,
      backward,
      'line'
    );
    const { x: scaledForwardX, y: scaledForwardY } = calculatePreviousNearestScaledPoint(
      scale,
      backward,
      forward,
      'line'
    );

    return (
      <g>
        <defs>
          <marker
            id="arrow"
            viewBox={`0 0 ${arrowDimensions.line} ${arrowDimensions.line}`}
            refX={arrowDimensions.line / 2}
            refY={arrowDimensions.line / 2}
            markerWidth="3"
            markerHeight="3"
            orient="auto-start-reverse"
            className={classes.arrow}
          >
            <path
              d={`M 0 0 L ${arrowDimensions.line} ${arrowDimensions.line / 2} L 0
              ${arrowDimensions.line} z`}
            />
          </marker>
        </defs>
        <line
          x1={scaledBackwardX}
          y1={scaledBackwardY}
          x2={scaledForwardX}
          y2={scaledForwardY}
          className={classNames(
            classes.bgSegment,
            disabled && classes.disabled,
            classes[correctness],
            className
          )}
          markerEnd="url(#arrow)"
          markerStart="url(#arrow)"
          {...rest}
        />
      </g>
    );
  }
}
