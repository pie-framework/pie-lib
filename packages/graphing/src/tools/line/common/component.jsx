import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles/index';
import { ToolPropType } from '../../types';
import debug from 'debug';
import { BasePoint, ArrowPoint } from '../../common/point';
import Segment from '../segment/segment';
import Vector from '../vector/vector';
import { point } from '../../../utils';
import classNames from 'classnames';
import { types } from '@pie-lib/plot';
import isEqual from 'lodash/isEqual';

const log = debug('pie-lib:graphing:segment');

const opacityPulsate = opacity => ({
  '0%': { opacity: '0.0' },
  '50%': { opacity },
  '100%': { opacity: '0.0' }
});

export const PointType = {
  x: PropTypes.number,
  y: PropTypes.number
};

const ComponentTypes = {
  segment: Segment,
  vector: Vector
};

export class RawBaseSegment extends React.Component {
  static propTypes = {
    building: PropTypes.bool,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    from: PropTypes.shape(PointType),
    to: PropTypes.shape(PointType),
    disabled: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    onDragStart: PropTypes.func,
    onDragStop: PropTypes.func,
    graphProps: types.GraphPropsType.isRequired
  };

  static defaultProps = {};

  constructor(props) {
    super(props);
    this.state = {
      draggedFrom: undefined
    };
  }

  moveFrom = from => {
    const { onChange, to } = this.props;
    const d = { from, to };

    this.setState({ draggedFrom: undefined }, () => {
      if (!isEqual(from, to)) {
        onChange(d);
      }
    });
  };

  dragFrom = draggedFrom => {
    log('[dragFrom] ', draggedFrom);
    this.setState({ draggedFrom });
  };

  moveTo = to => {
    const { onChange, from } = this.props;
    const d = { from, to };
    this.setState({ draggedTo: undefined }, () => {
      if (!isEqual(to, from)) {
        onChange(d);
      }
    });
  };

  dragTo = draggedTo => {
    this.setState({ draggedTo });
  };

  dragSegment = draggedFrom => {
    const { from, to } = this.props;

    log('dragSegment: ', draggedFrom, 'from: ', from, 'to: ', to);

    const diff = point(from).sub(point(draggedFrom));
    const draggedTo = point(to).sub(diff);

    this.setState({ draggedFrom, draggedTo, isSegmentDrag: true });
  };

  moveSegment = from => {
    const { onChange, from: propsFrom, to } = this.props;
    const diff = point(propsFrom).sub(point(from));
    const d = {
      from,
      to: point(to).sub(diff)
    };
    this.setState(
      {
        draggedFrom: undefined,
        draggedTo: undefined,
        isSegmentDrag: false
      },
      () => onChange(d)
    );
  };

  render() {
    let {
      from,
      disabled,
      classes,
      building,
      onDragStart,
      onDragStop,
      correctness,
      graphProps,
      type,
      to
    } = this.props;
    const { draggedFrom, draggedTo, isSegmentDrag } = this.state;
    log('[render] draggedFrom: ', draggedFrom, 'from: ', from);

    to = to || from;
    const f = draggedFrom || from;
    const s = draggedTo || to;
    const common = { graphProps };
    const Component = ComponentTypes[type];

    return (
      <g>
        <Component
          disabled={building || disabled}
          correctness={correctness}
          className={classNames(building && classes.segmentBuilding)}
          x={isSegmentDrag ? from.x : f.x}
          y={isSegmentDrag ? from.y : f.y}
          from={isSegmentDrag ? from : f}
          to={isSegmentDrag ? to : s}
          onDrag={this.dragSegment}
          onMove={this.moveSegment}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
        <BasePoint
          disabled={building || disabled}
          correctness={correctness}
          x={isSegmentDrag ? f.x : from.x}
          y={isSegmentDrag ? f.y : from.y}
          onMove={this.moveFrom}
          onDrag={this.dragFrom}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
        {type !== 'vector' && (
          <BasePoint
            disabled={building || disabled}
            correctness={correctness}
            x={isSegmentDrag ? s.x : to.x}
            y={isSegmentDrag ? s.y : to.y}
            onMove={this.moveTo}
            onDrag={this.dragTo}
            onDragStart={onDragStart}
            onDragStop={onDragStop}
            {...common}
          />
        )}
        {type !== 'segment' && (
          <ArrowPoint
            disabled={building || disabled}
            correctness={correctness}
            from={f}
            to={s}
            x={isSegmentDrag ? s.x : to.x}
            y={isSegmentDrag ? s.y : to.y}
            onMove={this.moveTo}
            onDrag={this.dragTo}
            onDragStart={onDragStart}
            onDragStop={onDragStop}
            {...common}
          />
        )}
      </g>
    );
  }
}

export const BaseSegment = withStyles(theme => ({
  outerLine: {
    fill: 'rgb(0,0,0,0)',
    stroke: theme.palette.primary.light,
    strokeWidth: 4,
    '&:hover': {
      strokeWidth: 6,
      stroke: theme.palette.primary.dark
    }
  },
  segmentBuilding: {
    stroke: theme.palette.secondary.light,
    animation: 'opacityPulse 2s ease-out',
    animationIterationCount: 'infinite',
    opacity: 1
  },
  '@keyframes opacityPulse': opacityPulsate('0.3')
}))(RawBaseSegment);

export default class Component extends React.Component {
  static propTypes = {
    ...ToolPropType,
    graphProps: types.GraphPropsType.isRequired
  };

  change = d => {
    const { onChange } = this.props;
    const m = {
      ...this.props.mark,
      from: d.from,
      to: d.to
    };
    onChange(this.props.mark, m);
  };

  render() {
    const { mark, onDragStart, onDragStop, graphProps } = this.props;
    return (
      <BaseSegment
        {...mark}
        onChange={this.change}
        onDragStart={onDragStart}
        onDragStop={onDragStop}
        graphProps={graphProps}
      />
    );
  }
}
