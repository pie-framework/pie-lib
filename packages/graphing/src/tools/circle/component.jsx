import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import debug from 'debug';
import { BasePoint } from '../shared/point';
import BgCircle from './bg-circle';
import { point } from '../../utils';
import classNames from 'classnames';
import { types } from '@pie-lib/plot';
import { rootEdgeComponent } from '../shared/line/with-root-edge';

const log = debug('pie-lib:graphing:circle');

const opacityPulsate = opacity => ({
  '0%': { opacity: '0.0' },
  '50%': { opacity },
  '100%': { opacity: '0.0' }
});

const getRadius = (root, outer) => {
  const c = point(root);
  return c.dist(point(outer));
};

export class RawBaseCircle extends React.Component {
  static propTypes = {
    building: PropTypes.bool,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    root: types.PointType,
    disabled: PropTypes.bool,
    edge: types.PointType,
    onChange: PropTypes.func.isRequired,
    onDragStart: PropTypes.func,
    onDragStop: PropTypes.func,
    graphProps: types.GraphPropsType.isRequired,
    onClick: PropTypes.func
  };

  static defaultProps = {
    onClick: () => ({})
  };

  dragRoot = root => {
    const { onChange } = this.props;
    const d = { root, edge: this.props.edge };

    onChange(d);
  };

  dragEdge = edge => {
    const { onChange } = this.props;
    const d = { root: this.props.root, edge };
    onChange(d);
  };

  dragCircle = root => {
    const { onChange } = this.props;
    const diff = point(this.props.root).sub(point(root));
    const edge = point(this.props.edge).sub(diff);
    const d = { root, edge };
    this.setState(
      {
        draggedroot: undefined,
        draggedOuter: undefined,
        isCircleDrag: false
      },
      () => {
        onChange(d);
      }
    );
  };

  render() {
    let {
      root,
      edge,
      disabled,
      classes,
      building,
      onDragStart,
      onDragStop,
      onClick,
      correctness,
      graphProps
    } = this.props;

    const common = { onDragStart, onDragStop, graphProps, onClick };

    edge = edge || root;

    const radius = getRadius(root, edge);

    return (
      <g>
        <BgCircle
          disabled={building || disabled}
          correctness={correctness}
          className={classNames(building && classes.bgCircleBuilding)}
          x={root.x}
          y={root.y}
          radius={radius}
          onDrag={this.dragCircle}
          {...common}
        />
        <BasePoint
          disabled={building || disabled}
          correctness={correctness}
          x={edge.x}
          y={edge.y}
          onDrag={this.dragEdge}
          {...common}
        />
        <BasePoint
          disabled={building || disabled}
          correctness={correctness}
          x={root.x}
          y={root.y}
          className={classes.root}
          onDrag={this.dragRoot}
          {...common}
        />
      </g>
    );
  }
}

export const BaseCircle = withStyles(theme => ({
  outerLine: {
    fill: 'rgb(0,0,0,0)',
    stroke: theme.palette.primary.light,
    strokeWidth: 4,
    '&:hover': {
      strokeWidth: 6,
      stroke: theme.palette.primary.dark
    }
  },
  root: {},
  bgCircleBuilding: {
    stroke: theme.palette.secondary.light,
    animation: 'opacityPulse 2s ease-out',
    animationIterationCount: 'infinite',
    opacity: 1
  },
  '@keyframes opacityPulse': opacityPulsate('0.3')
}))(RawBaseCircle);

const Component = rootEdgeComponent(BaseCircle);
export default Component;
