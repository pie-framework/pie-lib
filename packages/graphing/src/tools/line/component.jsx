import { lineToolComponent, lineBase, styles } from '../shared/line';
import React from 'react';
import PropTypes from 'prop-types';
import { trig, types } from '@pie-lib/plot';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import { ArrowMarker, genUid } from '../shared/arrow-head';
import { GraphPropsType } from '@pie-lib/plot/lib/types';
const markerId = genUid();

const lineStyles = theme => ({
  line: styles.line(theme),
  arrow: styles.arrow(theme)
});
export const ArrowedLine = props => {
  const { className, classes, correctness, disabled, graphProps, from, to, ...rest } = props;
  const { scale } = graphProps;
  const [eFrom, eTo] = trig.edges(graphProps.domain, graphProps.range)(from, to);
  return (
    <g>
      <defs>
        <ArrowMarker id={props.markerId || markerId} className={classes.arrow} />
      </defs>
      <line
        x1={scale.x(eFrom.x)}
        y1={scale.y(eFrom.y)}
        x2={scale.x(eTo.x)}
        y2={scale.y(eTo.y)}
        className={classNames(
          classes.line,
          disabled && classes.disabled,
          classes[correctness],
          className
        )}
        markerEnd={`url(#${props.markerId || markerId})`}
        markerStart={`url(#${props.markerId || markerId})`}
        {...rest}
      />
    </g>
  );
};

ArrowedLine.propTypes = {
  className: PropTypes.string,
  classes: PropTypes.object,
  correctness: PropTypes.string,
  disabled: PropTypes.bool,
  graphProps: GraphPropsType,
  from: types.PointType,
  to: types.PointType,
  markerId: PropTypes.string
};

const StyledArrowedLine = withStyles(lineStyles)(ArrowedLine);

const Line = lineBase(StyledArrowedLine);
const Component = lineToolComponent(Line);

export default Component;
