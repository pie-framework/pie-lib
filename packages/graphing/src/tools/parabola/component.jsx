import debug from 'debug';
import { buildDataPoints, parabolaFromTwoPoints } from '../utils';
import { withRootEdge, rootEdgeComponent } from '../shared/line/with-root-edge';

const log = debug('pie-lib:graphing:sine');

const Parabola = withRootEdge(props => {
  const { domain } = props.graphProps;

  const { root, edge } = props;
  const interval = 1;

  const dataPoints =
    edge && edge.x === root.x
      ? []
      : buildDataPoints(
          domain.min,
          domain.max,
          root,
          edge,
          interval,
          parabolaFromTwoPoints(root, edge)
        );
  log('dataPoints:', dataPoints);
  return { root: props.root, edge: props.edge, dataPoints };
});

const Component = rootEdgeComponent(Parabola);
export default Component;
