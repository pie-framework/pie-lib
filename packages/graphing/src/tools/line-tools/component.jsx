import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles/index';
import { ToolPropTypeFields } from '../types';
import debug from 'debug';
import { BasePoint, ArrowPoint } from '../common/point';
import { Label } from '../common/label';
import { point } from '../../utils';
import classNames from 'classnames';
import { types } from '@pie-lib/plot';
import isEqual from 'lodash/isEqual';
import { calculateThirdPointOnLine } from '../../utils';
import Components from './gridDraggableComponent';

const log = debug('pie-lib:graphing:segment');

const opacityPulsate = opacity => ({
  '0%': { opacity: '0.0' },
  '50%': { opacity },
  '100%': { opacity: '0.0' }
});

export const PointType = {
  x: PropTypes.number,
  y: PropTypes.number
};

export class RawBaseSegment extends React.Component {
  static propTypes = {
    building: PropTypes.bool,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    correctness: PropTypes.string,
    from: PropTypes.shape(PointType),
    to: PropTypes.shape(PointType),
    disabled: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    onDragStart: PropTypes.func,
    onDragStop: PropTypes.func,
    graphProps: types.GraphPropsType.isRequired,
    type: PropTypes.string,
    labelIsActive: PropTypes.bool
  };

  static defaultProps = {};

  constructor(props) {
    super(props);
    this.state = {
      draggedFrom: undefined,
      label: null
    };
  }

  moveFrom = from => {
    const { onChange, to } = this.props;
    const d = { from, to };

    this.setState({ draggedFrom: undefined }, () => {
      if (!isEqual(from, to)) {
        onChange(d);
      }
    });
  };

  dragFrom = draggedFrom => {
    this.setState({ draggedFrom });
  };

  moveTo = to => {
    const { onChange, from } = this.props;
    const d = { from, to };

    this.setState({ draggedTo: undefined }, () => {
      if (!isEqual(to, from)) {
        onChange(d);
      }
    });
  };

  dragTo = draggedTo => {
    this.setState({ draggedTo });
  };

  dragSegment = draggedFrom => {
    const { from, to } = this.props;
    const diff = point(from).sub(point(draggedFrom));
    const draggedTo = point(to).sub(diff);
    log('dragSegment: ', draggedFrom, 'from: ', from, 'to: ', to);

    this.setState({ draggedFrom, draggedTo, isSegmentDrag: true });
  };

  moveSegment = from => {
    const { onChange, from: propsFrom, to } = this.props;
    const diff = point(propsFrom).sub(point(from));
    const d = {
      from,
      to: point(to).sub(diff)
    };
    this.setState(
      {
        draggedFrom: undefined,
        draggedTo: undefined,
        isSegmentDrag: false,
        draggedToTest: null
      },
      () => onChange(d)
    );
  };

  getDirectionPoint = (lineStartsAt, lineEndsAt, rayPosition = 'forward') => {
    let arrow = lineEndsAt;
    let { from, graphProps } = this.props;
    const { draggedFrom, draggedTo, isSegmentDrag } = this.state;
    let arrowPoint;

    if (!isSegmentDrag) {
      if (rayPosition === 'forward') {
        arrowPoint = calculateThirdPointOnLine(lineEndsAt, lineStartsAt, graphProps);
      } else {
        arrowPoint = calculateThirdPointOnLine(lineStartsAt, lineEndsAt, graphProps);
      }

      if (isFinite(arrowPoint.x) && isFinite(arrowPoint.y)) {
        arrow = arrowPoint;
      }
    } else {
      const diff = point(from).sub(point(draggedFrom));
      if (rayPosition === 'forward') {
        arrowPoint = calculateThirdPointOnLine(draggedTo, draggedFrom, graphProps);
      } else {
        arrowPoint = calculateThirdPointOnLine(draggedFrom, draggedTo, graphProps);
      }

      arrow = point(arrowPoint).add(diff);
    }

    return arrow;
  };

  render() {
    let {
      from,
      disabled,
      classes,
      building,
      onDragStart,
      onDragStop,
      correctness,
      graphProps,
      type,
      to,
      labelIsActive
    } = this.props;
    const { draggedFrom, draggedTo, isSegmentDrag, label } = this.state;

    log('[render] draggedFrom: ', draggedFrom, 'from: ', from);

    to = to || from;
    const common = { graphProps };
    const Component = Components[type];

    const f = draggedFrom || from;
    const t = draggedTo || to;
    const startsAt = isSegmentDrag ? from : f;
    const endsAt = isSegmentDrag ? to : t;
    let forwardDirectionPoint = endsAt;
    let backwardDirectionPoint = startsAt;

    if (type === 'ray' || type === 'line') {
      forwardDirectionPoint = this.getDirectionPoint(startsAt, endsAt, 'backward');
    }

    if (type === 'line') {
      backwardDirectionPoint = this.getDirectionPoint(startsAt, endsAt, 'forward');
    }

    return (
      <g
        onClick={() => {
          if (labelIsActive) {
            this.setState({ label: '' });
          }
        }}
      >
        {label !== null && (
          <Label
            disabled={building || disabled}
            correctness={correctness}
            onChange={value => this.setState({ label: value })}
            onRemove={() => this.setState({ label: null })}
            x={isSegmentDrag ? f.x : from.x}
            y={isSegmentDrag ? f.y : from.y}
            {...common}
          />
        )}
        <Component
          disabled={building || disabled}
          correctness={correctness}
          className={classNames(building && classes.segmentBuilding)}
          x={startsAt.x}
          y={startsAt.y}
          from={startsAt}
          to={endsAt}
          backward={backwardDirectionPoint}
          forward={forwardDirectionPoint}
          onDrag={this.dragSegment}
          onMove={this.moveSegment}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
        <BasePoint
          disabled={building || disabled}
          correctness={correctness}
          x={isSegmentDrag ? f.x : from.x}
          y={isSegmentDrag ? f.y : from.y}
          onMove={this.moveFrom}
          onDrag={this.dragFrom}
          onDragStart={onDragStart}
          onDragStop={onDragStop}
          {...common}
        />
        {type !== 'vector' && (
          <BasePoint
            disabled={building || disabled}
            correctness={correctness}
            x={isSegmentDrag ? t.x : to.x}
            y={isSegmentDrag ? t.y : to.y}
            onMove={this.moveTo}
            onDrag={this.dragTo}
            onDragStart={onDragStart}
            onDragStop={onDragStop}
            {...common}
          />
        )}
        {type === 'vector' && (
          <ArrowPoint
            disabled={building || disabled}
            correctness={correctness}
            from={f}
            to={t}
            x={isSegmentDrag ? t.x : to.x}
            y={isSegmentDrag ? t.y : to.y}
            onMove={this.moveTo}
            onDrag={this.dragTo}
            onDragStart={onDragStart}
            onDragStop={onDragStop}
            {...common}
          />
        )}
      </g>
    );
  }
}

export const BaseSegment = withStyles(theme => ({
  outerLine: {
    fill: 'rgb(0,0,0,0)',
    stroke: theme.palette.primary.light,
    strokeWidth: 4,
    '&:hover': {
      strokeWidth: 6,
      stroke: theme.palette.primary.dark
    }
  },
  segmentBuilding: {
    stroke: theme.palette.secondary.light,
    animation: 'opacityPulse 2s ease-out',
    animationIterationCount: 'infinite',
    opacity: 1
  },
  '@keyframes opacityPulse': opacityPulsate('0.3')
}))(RawBaseSegment);

export default class Component extends React.Component {
  static propTypes = {
    ...ToolPropTypeFields,
    graphProps: types.GraphPropsType.isRequired
  };

  change = d => {
    const { onChange } = this.props;
    const m = {
      ...this.props.mark,
      from: d.from,
      to: d.to
    };
    onChange(this.props.mark, m);
  };

  render() {
    const { mark, onDragStart, onDragStop, graphProps, labelIsActive } = this.props;
    return (
      <BaseSegment
        {...mark}
        onChange={this.change}
        onDragStart={onDragStart}
        onDragStop={onDragStop}
        graphProps={graphProps}
        labelIsActive={labelIsActive}
      />
    );
  }
}
