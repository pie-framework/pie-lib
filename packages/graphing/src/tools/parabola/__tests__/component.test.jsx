import { shallow } from 'enzyme';
import React from 'react';
import { withRootEdge } from '../../shared/line/with-root-edge';
import { buildDataPoints, parabolaFromTwoPoints } from '../../utils';
import { utils } from '@pie-lib/plot';

import { graphProps as getGraphProps } from '../../../__tests__/utils';

const { xy } = utils;
jest.mock('../../utils', () => ({
  buildDataPoints: jest.fn().mockReturnValue([]),
  parabolaFromTwoPoints: jest.fn(() => jest.fn())
}));

jest.mock('../../shared/line/with-root-edge', () => ({
  withRootEdge: jest.fn(),
  rootEdgeComponent: jest.fn()
}));
describe('Parabola', () => {
  let fnBody;
  let graphProps;
  let root;
  let edge;
  let result;
  beforeEach(() => {
    require('../component');
    fnBody = withRootEdge.mock.calls[0][0];
    graphProps = getGraphProps();
    root = xy(0, 0);
    edge = xy(1, 1);

    result = fnBody({ graphProps, root, edge });
  });
  it('fnBody is not null', () => {
    expect(fnBody).toBeDefined();
  });

  it('calls buildDataPoints', () => {
    const { domain } = graphProps;
    expect(buildDataPoints).toHaveBeenCalledWith(
      domain.min,
      domain.max,
      root,
      edge,
      1,
      expect.anything()
    );
  });
  it('calls parabolaFromTwoPoints', () => {
    expect(parabolaFromTwoPoints).toHaveBeenCalledWith(root, edge);
  });

  it('returns dataPoints', () => {
    expect(result).toEqual({ root, edge, dataPoints: [] });
  });
});
