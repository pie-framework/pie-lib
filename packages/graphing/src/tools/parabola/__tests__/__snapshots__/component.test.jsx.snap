// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshot renders 1`] = `
<RootEdge
  edge={
    Object {
      "x": 2,
      "y": 3,
    }
  }
  graphProps={
    Object {
      "domain": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "range": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "scale": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
      "size": Object {
        "height": 400,
        "width": 400,
      },
      "snap": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
    }
  }
  onChange={[Function]}
  onClick={[Function]}
  root={
    Object {
      "x": 0,
      "y": 0,
    }
  }
/>
`;
